rectools/visuals/metrics_app.py

```diff
@@ -85,125 +97,267 @@ def construct(
-    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:  # pragma: no cover
-        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+    def _make_chart_data_fold(self, fold_number: int) -> pd.DataFrame:
+        return self.data[self.data[Columns.Split] == fold_number].reset_index(drop=True)
 
     @lru_cache
     def _make_chart_data_avg(self) -> pd.DataFrame:
-        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+        metric_data_columns = [Columns.Model] + self.metric_names
+        meta_data_columns = [Columns.Model] + self.meta_names
+        # separate metric data because meta data could contain strings
+        metrics_data = self.data[metric_data_columns].groupby([Columns.Model], sort=False).mean().reset_index()
+        meta_data = self.data[meta_data_columns].drop_duplicates()
+        return metrics_data.merge(meta_data, on=Columns.Model, how="left").reset_index(drop=True)
+
+    @staticmethod
+    @lru_cache
+    def _split_to_meta_and_model(raw_string: str, splitter: str = ", ") -> tp.Tuple[str, str]:
+        splitted_row = raw_string.split(splitter, 1)
+        if len(splitted_row) > 1:
+            meta_value, model_name = splitted_row
+            return meta_value, model_name
+        return "", raw_string
+
+    def _create_chart_figure(
+        self,
+        data: pd.DataFrame,
+        metric_x: str,
+        metric_y: str,
+        color: str,
+        legend_title: str,
+    ) -> go.Figure:  # pragma: no cover
+        scatter_kwargs = {
+            "width": WIDGET_WIDTH,
+            "height": WIDGET_HEIGHT,
+        }
+        scatter_kwargs.update(self.scatter_kwargs)
+
+        data = data.sort_values(by=color, ascending=True)
+        data[color] = data[color].astype(str)  # to treat colors values as categorical
+
+        fig = px.scatter(
+            data,
+            x=metric_x,
+            y=metric_y,
+            color=color,
+            symbol=Columns.Model,
+            **scatter_kwargs,
+        )
+        fig.update_layout(margin={"t": TOP_CHART_MARGIN}, legend_title=legend_title, showlegend=self.show_legend)
+        fig.update_coloraxes(showscale=False)
+        return fig
 
-    def _update_chart(
+    def _update_figure_widget(
         self,
         fig_widget: go.FigureWidget,
         metric_x: widgets.Dropdown,
         metric_y: widgets.Dropdown,
         use_avg: widgets.Checkbox,
         fold_i: widgets.Dropdown,
-    ) -> None:
-        data = self._make_chart_data_avg() if use_avg.value else self._make_chart_data(fold_i.value)
-        existing_traces = {trace.name: trace for trace in fig_widget.data}
+        meta_feature: widgets.Dropdown,
+        use_meta: widgets.Checkbox,
+    ) -> None:  # pragma: no cover
+        chart_data = self._create_chart_data(use_avg, fold_i)
+        color_clmn = meta_feature.value if use_meta.value else Columns.Model
```

Comment:
Please use `col` or `column`, it's kinda stylecode :)

Reply:
Done